### DISCUSSION TITLE
How to resolve the issue of text not displaying fully (misalignment) after being written back.

### DISCUSSION THREAD

**Discussion by xiaolibuzai-ovo** [pymupdf/PyMuPDF#4122] [Looking for help]
The right side shows the original file, while the left side shows the effect after writing back the text.
<img width="1501" alt="image" src="https://github.com/user-attachments/assets/383602ff-c5a6-4fbf-ac30-2885af644492">

code:
```python

import pymupdf
import json

docs = pymupdf.open("jiacu.pdf")  # any supported document type
for page in docs:
    pages.append(page.get_text("dict",flags=pymupdf.TEXTFLAGS_TEXT))
    for link in page.get_links():
        page.delete_link(link)
    blocks=page.get_text("dict",flags=pymupdf.TEXTFLAGS_TEXT)["blocks"]
    for block in blocks:
        lines = block['lines']
        for line in lines:
            spans = line['spans']
            dir = line['dir']
            for i in range(len(spans)):
                span = spans[i]
                bbox = span['bbox']
                text = span["text"]
                font = span['font']
                size = span['size']
                color = "%02x%02x%02x" % pymupdf.sRGB_to_rgb(span["color"])
                WHITE = pymupdf.pdfcolor["white"]
                page.draw_rect(bbox, color=None, fill=WHITE)
                text = text.replace('<',"《")
                text = text.replace('>',"》")
                is_italic = bool(span['flags'] & 2**1)        
                is_serifed = bool(span['flags'] & 2**2)       
                is_monospaced = bool(span['flags'] & 2**3)    
                is_bold = bool(span['flags'] & 2**4)           
                rotate=0
                if dir==(0.0,-1.0):
                    rotate=90
                css = f"""
                * {{
                    font-family: '{font}';
                    font-size: {size}px;
                    color: #{color};
                    padding: 0;   
                    margin: 0;    
                    {'font-style: italic;' if is_italic else ''}
                    {'font-weight: bold;' if is_bold else ''}
                    {'font-family: serif;' if is_serifed else ''}
                    {'font-family: monospace;' if is_monospaced else ''}
                }}
                """                
                page.insert_htmlbox(bbox, text,css=css, rotate=rotate)

docs.ez_save("jiacu.pdf-new.pdf",garbage=4)


```

**Comment by xiaolibuzai-ovo** [comment]
file:
[Why Not Transform Chat Large Language Models to Non-English_.pdf](https://github.com/user-attachments/files/18047626/Why.Not.Transform.Chat.Large.Language.Models.to.Non-English_.pdf)

**Comment by JorjMcKie** [comment]
Method `page.insert_htmlbox` is based on the `Story` class. As such, layout details are fully under control of this logic and cannot be influenced.

**Reply by xiaolibuzai-ovo** [reply]
Thank you very much for your reply. I apologize for being a beginner with PyMuPDF. Could you please provide an example of how to use 'story'? I couldn't seem to find it in the documentation.

**Reply by JorjMcKie** [reply]
There is [this](https://pymupdf.readthedocs.io/en/latest/recipes-stories.html) chapter in the documentation dedicated to story.

**Reply by xiaolibuzai-ovo** [reply]
thanks!

**Reply by xiaolibuzai-ovo** [reply]
> Method `page.insert_htmlbox` is based on the `Story` class. As such, layout details are fully under control of this logic and cannot be influenced.

<img width="1480" alt="image" src="https://github.com/user-attachments/assets/c74fc8ed-aa82-41f7-b2ef-7ad42189c288">
I have marked the original file and the text I wrote back with red boxes. I am curious why the text on the right side is not fully displayed, even though the boxes are the same size. For example, in the area marked with the black box

**Reply by Mahendranath94** [reply]
Hi @xiaolibuzai-ovo Were you able to fix the above issue?
